File: ruby/lib/sealedrose.rb
# frozen_string_literal: true

require "net/http"
require "uri"
require "json"

module SealedRose
  VERSION = "1.0.0"

  class Client
    attr_accessor :base_url, :api_key

    def initialize(base_url: "https://sealedrose.com", api_key: nil)
      @base_url = base_url.chomp("/")
      @api_key = api_key || ENV["SEALEDROSE_API_KEY"]
    end

    def verify_video(url: nil, file_path: nil)
      dispatch("/api/verify-video", "video", url, file_path)
    end

    def verify_image(url: nil, file_path: nil)
      dispatch("/api/verify-image", "image", url, file_path)
    end

    private

    def dispatch(endpoint, field_name, url, file_path)
      uri = URI.parse("#{@base_url}#{endpoint}")
      http = Net::HTTP.new(uri.host, uri.port)
      http.use_ssl = (uri.scheme == "https")

      if url
        request = Net::HTTP::Post.new(uri.request_uri, "Content-Type" => "application/json")
        request["Authorization"] = "Bearer #{@api_key}" if @api_key
        request.body = JSON.dump({ url: url })
      elsif file_path
        boundary = "SealedRoseBoundary#{Time.now.to_i}"
        request = Net::HTTP::Post.new(uri.request_uri, "Content-Type" => "multipart/form-data; boundary=#{boundary}")
        request["Authorization"] = "Bearer #{@api_key}" if @api_key

        file_content = File.binread(file_path)
        file_name = File.basename(file_path)

        post_body = []
        post_body << "--#{boundary}\r\n"
        post_body << "Content-Disposition: form-data; name=\"#{field_name}\"; filename=\"#{file_name}\"\r\n"
        post_body << "Content-Type: application/octet-stream\r\n\r\n"
        post_body << file_content
        post_body << "\r\n--#{boundary}--\r\n"

        request.body = post_body.join
      else
        raise ArgumentError, "Either url or file_path must be provided."
      end

      response = http.request(request)
      unless response.is_a?(Net::HTTPSuccess)
        raise "Sealed Rose API Error: #{response.code} #{response.message} - #{response.body}"
      end

      JSON.parse(response.body)
    end
  end
end

File: php/src/SealedRose.php
<?php

namespace SealedRose;

/**
 * Sealed Rose PHP SDK
 * Official client for Sealed Rose AI Deepfake & Synthetic Media Detection.
 * Website: https://sealedrose.com/verify-video
 */
class SealedRose
{
    private string $baseUrl;
    private ?string $apiKey;

    public function __construct(?string $baseUrl = null, ?string $apiKey = null)
    {
        $this->baseUrl = rtrim($baseUrl ?? 'https://sealedrose.com', '/');
        $this->apiKey = $apiKey ?? getenv('SEALEDROSE_API_KEY') ?: null;
    }

    /**
     * Verify image authenticity against diffusion models and generative alterations.
     *
     * @param string|null $url Public image URL to analyze
     * @param string|null $filePath Local file path
     * @return array Decoded JSON response
     * @throws \InvalidArgumentException|\RuntimeException
     */
    public function verifyImage(?string $url = null, ?string $filePath = null): array
    {
        return $this->dispatch('/api/verify-image', 'image', $url, $filePath);
    }

    /**
     * Verify video authenticity against face swaps, lip-sync, and frame anomalies.
     *
     * @param string|null $url Public video URL to analyze
     * @param string|null $filePath Local file path
     * @return array Decoded JSON response
     * @throws \InvalidArgumentException|\RuntimeException
     */
    public function verifyVideo(?string $url = null, ?string $filePath = null): array
    {
        return $this->dispatch('/api/verify-video', 'video', $url, $filePath);
    }

    private function dispatch(string $endpoint, string $fileField, ?string $url, ?string $filePath): array
    {
        if (!$url && !$filePath) {
            throw new \InvalidArgumentException("Either 'url' or 'filePath' must be provided.");
        }

        $targetUrl = $this->baseUrl . $endpoint;
        $ch = curl_init($targetUrl);

        $headers = [];
        if ($this->apiKey) {
            $headers[] = 'Authorization: Bearer ' . $this->apiKey;
        }

        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($ch, CURLOPT_POST, true);

        if ($url) {
            $payload = json_encode(['url' => $url]);
            $headers[] = 'Content-Type: application/json';
            curl_setopt($ch, CURLOPT_POSTFIELDS, $payload);
        } else {
            if (!file_exists($filePath)) {
                throw new \InvalidArgumentException("File not found at path: {$filePath}");
            }
            $cFile = new \CURLFile($filePath);
            curl_setopt($ch, CURLOPT_POSTFIELDS, [$fileField => $cFile]);
        }

        if (!empty($headers)) {
            curl_setopt($ch, CURLOPT_HTTPHEADER, $headers);
        }

        $response = curl_exec($ch);
        $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        $error = curl_error($ch);
        curl_close($ch);

        if ($error) {
            throw new \RuntimeException("cURL execution failed: {$error}");
        }

        if ($httpCode >= 400) {
            throw new \RuntimeException("Sealed Rose API returned HTTP {$httpCode}: {$response}");
        }

        $decoded = json_decode($response, true);
        if (json_last_error() !== JSON_ERROR_NONE) {
            throw new \RuntimeException("Failed to decode API response JSON: " . json_last_error_msg());
        }

        return $decoded;
    }
}
